### PR TITLE
make adjoint emitter find correct adjoint to emit; fix two activity analysis bugs

### PR DESCRIPTION
**Pull request by marcrasi**
Fixes `simple_mul` by making the adjoint emitter find adjoint values using original function params.

Fixes `chain_rule` by making values active when they vary for any parameter. (Previously, values had to vary for all parameters.)

Also I noticed a bug where varied values were being collected incorrectly because `visitedVariedValues` didn't get cleared for new searches. A gradient wrt a subset of the function's parameters should be able to expose this bug, but when I try to write one it crashes somewhere else (the gradient SILFunction tries to return a tuple with too many elements). So no test for this bug, for now.

### CODE REVIEW

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 1]
```diff
@@ -1288,8 +1288,8 @@ class DifferentiableActivityInfo {
                 unsigned independentVariableIndex) const;
   bool isUseful(SILValue value,
                 unsigned dependentVariableIndex) const;
-  bool isVaried(SILValue value,
-                const llvm::SmallBitVector &parameterIndices) const;
+  bool isVariedByAny(SILValue value,
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: The semantics of `isVaried` is already "is varied for any parameter". So the name need not be changed.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 2]
```diff
@@ -2924,12 +2925,15 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
...
+    for (auto paramPair : zip(original.getArgumentsWithoutIndirectResults(),
+                              originalParametersInAdj)) {
+      auto origParam = std::get<0>(paramPair);
+      auto adjParam = std::get<1>(paramPair);
+      auto adjVal = getAdjointValue(origParam);
+      if (adjParam->getType().isObject())
         retElts.push_back(materializeAdjoint(adjVal, adjLoc));
       else
-        materializeAdjointIndirect(adjVal, param);
+        materializeAdjointIndirect(adjVal, adjParam);
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Really great fix!

**Review by rxwei — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**